### PR TITLE
FEATURE: Add campus [ON-69] [ON-78]

### PR DESCRIPTION
**Pull request by AkshatBajaj**
# Description

This PR introduces the campus....

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation if appropriate
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have created or extended unit tests to address my new additions
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

If you have any questions, please contact @macite or @jakerenzella.

### CODE REVIEW

File: app/api/projects_api.rb [hunk 1]
```diff
@@ -85,13 +86,28 @@ class ProjectsApi < Grape::API
...
         elsif tutorial_id == -1
           project.tutorial = nil
-          project.save!
         else
           error!({ error: "Couldn't find Tutorial with id=#{params[:tutorial_id]}" }, 403)
         end
+        project.save!
+
+        # If we are only updating the campus
+      elsif !params[:campus_id].nil?
```
Comment (macite) [inline app/api/projects_api.rb]: Use .present

File: app/api/projects_api.rb [hunk 2]
```diff
@@ -85,13 +86,28 @@ class ProjectsApi < Grape::API
...
+        unless authorise? current_user, project, :change_campus
```
Comment (macite) [inline app/api/projects_api.rb]: Make sure only admin, convenor, and tutor can change project campus

File: app/models/campus.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
+class Campus < ActiveRecord::Base
+  # Relationships
+  has_many    :tutorials
+  has_many    :projects
+
+  # Callbacks - methods called are private
+  before_destroy :can_destroy?
+
+  validates :name, presence: true
```
Comment (macite) [inline app/models/campus.rb]: unique

File: app/models/campus.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  validates :mode, presence: true
+  validates :abbreviation, presence: true
```
Comment (macite) [inline app/models/campus.rb]: unique

File: app/models/unit.rb [hunk 5]
```diff
@@ -388,14 +389,15 @@ def enrol_student(user, tutorial = nil)
...
-    project = Project.create!(
+    project = Project.new(
       user_id: user.id,
       unit_id: id,
       task_stats: '0.0|1.0|0.0|0.0|0.0'
     )
 
     project.tutorial_id = tutorial_id unless tutorial_id.nil?
-    project.save
+    project.campus = campus
```
Comment (macite) [inline app/models/unit.rb]: move up into new ( ... )

File: app/models/unit.rb [hunk 6]
```diff
@@ -678,13 +682,14 @@ def update_student_enrolments(changes, import_settings, result)
 
           # Add the user to the project (if not already in there)
           if user_project.nil?
+            campus = Campus.find_by(name: campus_name)
```
Comment (macite) [inline app/models/unit.rb]: Also try by abbreviation

Comment (macite) [reply app/models/unit.rb]: So if not the abbreviation then try name

File: app/models/unit.rb [hunk 7]
```diff
@@ -961,15 +968,18 @@ def import_groups_from_csv(group_set, file)
           tutorial = tutorial_with_abbr(tutorial_abbr)
           if tutorial.nil?
             change += 'Created new tutorial. '
+            campus = Campus.find_by(name: campus_name)
```
Comment (macite) [inline app/models/unit.rb]: Maybe add a general find_by_data?? to find by abbr then by name if needed

Comment (macite) [reply app/models/unit.rb]: In Campus model

File: lib/tasks/populate.rake [hunk 8]
```diff
@@ -7,7 +7,7 @@ namespace :db do
     tutes = unit.tutorials
     for student_count in 0..2000
       student = find_or_create_student("student_#{student_count}")
-      proj = unit.enrol_student(student, tutes[student_count % tutes.count])
+      proj = unit.enrol_student(student, Campus.first, tutes[student_count % tutes.count])
```
Comment (macite) [inline lib/tasks/populate.rake]: Make sure campus matches tutorial

File: test/factories/units_factory.rb [hunk 9]
```diff
@@ -10,6 +10,12 @@
     unit
   end
 
+  factory :campus do
+    name "Melbourne"
```
Comment (macite) [inline test/factories/units_factory.rb]: Have a look at expanding this to have multiple campuses

File: app/api/projects_api.rb (hunk 2 above)
Comment (AkshatBajaj) [reply app/api/projects_api.rb]: https://github.com/AkshatBajaj/doubtfire-api/blob/7583cdbb796dd5bf838100f2d296cc4cec674dd0/app/models/project.rb#L58

**Pull request merged**